GH-2924: Lateral - fixed injection for tables and enhanced QueryExec API for easier testing.

Aklakan opened the pull request:
GitHub issue resolved #2924

Pull request Description:

* Added a new test case with nested laterals.
* Fixed nested laterals by merging compatible bindings in `QueryIterLateral`
* While writing the tests, I added a few shorthands to the QueryExecBuilder API: Its now possible to write `QueryExec.dataset(...).query("SELECT ...").table()` to generally safely obtain a detached materialized in-memory table without having to worry about resource leaks or try-with-resources blocks. In the revised tests, the tables are used directly when asserting the equality of expected and actual results. In an attempt to allow for cleanly detaching BindingTDB instances, I added `Binding.detach()` with appropriate implementations. `Binding.detach` avoids copying of bindings that are already in-memory, and is thus more subtle than `BindingFactory.copy`.
* Added a `TableBuilder` class to build immutable tables.

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

Aklakan commented on jena-arq/src/main/java/org/apache/jena/query/Query.java:
```diff
@@ -98,7 +98,7 @@ public class Query extends Prologue implements Cloneable, Printable
 
     // VALUES trailing clause
-    protected TableData valuesDataBlock = null ;
+    protected TableN valuesDataBlock = null ;
```
Should the valuesDataBlock be mutable or not? One test case written by me some years ago assumed it to be mutable - but the implementation of TableData suggests to be immutable as the addBinding method is overridden to reject updates.

As I see it: TableN -> mutable, TableData -> immutable - much like guava's List and ImmutableList.

Aklakan replied:
Ok, so I'll change back to TableData and update the test case to expect failure when trying to modify the rows of a queries valuesDataBlock directly.

Aklakan replied:
Query has been reverted and the test case has been updated.

Aklakan commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterLateral.java:
```diff
@@ -292,17 +307,33 @@ public Op transform(OpTable opTable) {
-                builder.reset();
-                builder.addAll(row);
-                builder.addAll(binding);
-                table2.addBinding(builder.build());
```
addBinding checks whether the table's variables need to be updated; this check is not needed here.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/algebra/table/TableData.java:
```diff
@@ -18,23 +18,21 @@
 public class TableData extends TableN {
     public TableData(List<Var> variables, List<Binding> rows) {
-        super(variables, rows) ;
+        super(Collections.unmodifiableList(variables), Collections.unmodifiableList(rows)) ;
```
`Collections.unmodifiableList` only puts a wrapper around the base list. The base `variables` and `rows` can be modified.

Aklakan replied:
The constructor already existed. The wrappers prevent mutation via `TableN.getRows().add()` which wasn't the case before. Its not perfect, but from your comment

> It should be "immutable once built"

its a step closer towards `TableData` being immutable.

afs commented on jena-arq/src/main/java/org/apache/jena/query/ResultSet.java:
```diff
@@ -103,4 +103,8 @@ public default ResultSet materialise() {
 
     public void close();
+
+    default RowSet toRowSet() {
```
I prefer `asRowSet` -- `as*` when it is the  same thing, different presentation, and `to*` when it is a created conversion.

c.f. `Dataset::asDatasetGraph`

Aklakan replied:
Renamed to `asRowSet`.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/algebra/TableFactory.java:
```diff
@@ -27,31 +28,45 @@
     { return new Table1(var, value) ; }
+
+    /** Creates a mutable table from the detached bindings of the row set. */
+    public static Table create(RowSet rs)
```
I don't understand what going on. Could you explain this please? 

This seems to duplicate the intent of `RowSet.materialize` and `RowSet.rewindable`.

Aklakan replied:
The difference (to my understanding) is that a Table acts as a Collection and a RowSet as an Iterator.
A Table can thus be seen as a factory for RowSets via `Table.toRowSet()` - each obtained RowSet is an independent iterator over the Table.
My qualms with RowSet.materialize/rewindable is that these methods operate on the iterator level; it seemed clean to me to add Table as a collection-level view over a list of bindings.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/binding/BindingBase.java:
```diff
@@ -202,4 +202,19 @@ public static int hashCode(Binding bind) {
+
+    @Override
+    public Binding detach() {
+        Binding newParent = parent == null ? null : parent.detach();
+        Binding result = newParent == parent
+                ? detachWithOriginalParent()
+                : detachWithNewParent(newParent);
```
```suggestion
        Binding newParent = (parent == null) ? null : parent.detach();
        Binding result = (newParent == parent)
                ? detachWithOriginalParent()
                : detachWithNewParent(newParent);
```

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/binding/BindingBase.java:
```diff
@@ -202,4 +202,19 @@ public static int hashCode(Binding bind) {
+        return this;
+    }
+
+    protected abstract Binding detachWithNewParent(Binding newParent);
```
This would be nicer in `BindingFactory` ... if we were using Java21 and could use switch patterns. Oh well, another time.

Aklakan replied:
I was also thinking that BindingFactory could take care of making bindings independent. But then again, why should BindingFactory have special logic for BindingTDB instances? In principle, `BindingTDB.detach` could be implemented as just setting filling the already present (but unused) cache - so no copy would be needed then.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/exec/QueryExec.java:
```diff
@@ -54,6 +54,12 @@ public static QueryExecBuilder graph(Graph graph) {
     }
 
+    /** Create a {@link QueryExecBuilder} initialized with an empty dataset. */
+    public static QueryExecBuilder emptyDataset() {
```
This is only used from `TestQueryExecution`. Please can the code go there so it is not in the runtime API.? 

It can use`DatasetGraphFactory.empty()`.

```java
return QueryExec.dataset(DatasetGraphFactory.empty()`);
```

Aklakan replied:
It was just a suggestion for a shortcut - I'll remove it.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/exec/QueryExecBuilder.java:
```diff
@@ -106,4 +116,18 @@ public default boolean ask() {
+     * Creates and returns an independent in-memory table by materializing the underlying row set.
+     * Subsequently, {@link Table#toRowSet()} can be used to obtain a fresh row set view over the table.
+     */
+    public default Table table() {
```
See the question about "RowSet.matrialize/rewindable" above.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/exec/RowSet.java:
```diff
@@ -93,4 +93,8 @@ public default Stream<Binding> stream() {
         return RowSetStream.create(vars, qIter);
     }
+
+    default ResultSet toResultSet() {
```
```suggestion
    default ResultSet asResultSet() {
```

afs commented on jena-arq/src/test/java/org/apache/jena/sparql/graph/GraphsTests.java:
```diff
@@ -106,56 +110,80 @@ protected void fillDataset(Dataset dataset) {
+        } finally {
+            ds.close() ;
+        }
+        assertEquals(expected, actual) ;
```
See `ResultSetCompare.exactEquals` (which should probably be in `RowSetOps`).

Aklakan replied:
As mentioned, I view Table as a collection and RowSet as an Iterator.
If I am not mistaken, then table's `equals()` method naturally and succinctly performs an exact comparison (using term equality). For working with iterators (RowSet/ResultSet) one should of course use ResultSetCompare / RowSetOps methods.
In this case, one could also write a bit more verbose: `assertTrue(ResultSetCompare.exactEquals(expected.toRowSet(), actual.toRowSet());`

afs commented:
First part of a review - trying to understand the machinery changes. There seem to be a lot of them and I'm trying to understand what are necessary.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterLateral.java:
```diff
@@ -292,17 +295,33 @@ public Op transform(OpTable opTable) {
+            // Index variables in a set if there are more than a few of them.
+            Collection<Var> tableVarsIndex = tableVars.size() > 4 ? new HashSet<>(tableVars) : tableVars;
+            binding.vars().forEachRemaining(v -> {
+                if (tableVarsIndex.contains(v)) {
```
The table vars can't contain a binding var. That case is caught by the 

https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md  
"Disallow syntactic forms that set variables that may already be present in the current row."

It's tested in `SyntaxVarScope.checkLATEAL`.

(I removed the `if`, left the `else` and the second test passes). 

If so, it simplifies to `vars.add(v);`
And hence "vars = tableVars" (copy not needed).

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterLateral.java:
```diff
@@ -292,17 +295,33 @@ public Op transform(OpTable opTable) {
-                builder.addAll(row);
-                builder.addAll(binding);
-                table2.addBinding(builder.build());
+                if (Algebra.compatible(row, binding, commonVars.iterator())) {
```
If there is a variable in common from substitution it must be the same term.

So this is a contains relationship. (Given the comments above there may be a simpler way.)

Aklakan replied:
There is potentially another can of worms I was about to open w.r.t. sep7:

Perhaps its worthy of opening an issue and discussion on sparql-dev.  The reason I added the compatibility check was because I had the following query in mind, which for reasons I do not understand is currently prohibited:

```sparql
SELECT * { VALUES ?department {
    <urn:dept1>
    <urn:dept2>
  }
  LATERAL { SELECT * { VALUES (?department ?employee) {
        ( <urn:dept1> <urn:person1> )
        ( <urn:dept1> <urn:person2> )
        ( <urn:dept2> <urn:person1> )
      }
    } ORDER BY ?employee LIMIT 1
  }
}
```

| department  | employee      |
|--------------------|---------|
| <urn:dept1> | <urn:person1> |
| <urn:dept2> | <urn:person1> |
-------------------------------

Removing the restrictions `SyntaxVarScope.checkLATERAL` on `Element_Bind` and `Element_Data`) and adding the check for compatible bindings (as done in the PR) naturally makes the query work.

afs replied:
There must be a problem with `SyntaxVarScope.checkLATERAL` because the `SELECT *` is effectively `SELECT ?employee` which works.

Aklakan replied:
> `SELECT *` is effectively `SELECT ?employee` which works.

Hm, depending on the viewpoint it is the effective query *after* substitution, but before that, its `SELECT ?department ?employee` as to make `?department` in-scope such that the substitution could affect the table and discard its incompatible rows.

I think the scope error is in-line with the SEP7 statement:

> "Disallow syntactic forms that set variables that may already be present in the current row."

But I think that this is too restrictive, because it prevents inlining data using VALUES blocks. IMO if substitution leads to incompatible bindings then they should just be discarded as usual - no? Perhaps you have corner cases in mind where under this logic the substitution could become ambiguous and that this might be the reason for the restrictions.

Removing the restrictions on `Element_Bind` and `Element_Data` restrictions in `checkLATERAL` in order to make the above "department-employee" query work causes the following test cases to fail:

 syntax-lateral-bad-{[04](https://github.com/apache/jena/blob/main/jena-arq/testing/ARQ/Syntax/Syntax-Lateral/syntax-lateral-bad-04), [05](https://github.com/apache/jena/blob/main/jena-arq/testing/ARQ/Syntax/Syntax-Lateral/syntax-lateral-bad-05), [06](https://github.com/apache/jena/blob/main/jena-arq/testing/ARQ/Syntax/Syntax-Lateral/syntax-lateral-bad-06), [08](https://github.com/apache/jena/blob/main/jena-arq/testing/ARQ/Syntax/Syntax-Lateral/syntax-lateral-bad-08)}.arq

For example, `syntax-lateral-bad-04`:
```sparql
SELECT * {
   ?s ?p ?o
   LATERAL { BIND( 123 AS ?o) }
}
```

This is certainly not an efficient way to write a query, but I don't understand why the query should be syntactically illegal in the first place. To me it looks like it would only retain bindings where ?o = 123. So it would be the task of an optimizer to rewrite the `Lateral/Bind` to a plain `FILTER(?o = 123)`.

When disabling the SyntaxVarScope check for ElementBind, each ?o gets injected into the unit table beneath `(extend ((?o 123))  (table unit))` such as `(extend ((?o 123))  (table (row (?o 456))))` and the evaluation of OpExtend discards the incompatible binding - which seems like the natural way it should be to me.

afs replied:
> Hm, depending on the viewpoint it is the effective query _after_ substitution,

My bad - I was thinking about `GROUP BY`.

For `ORDER BY`, the query is illegal by the assignment restriction.

> To me it looks like it would only retain bindings where ?o = 123.

That would turn `BIND` into a filter. (ARQ's extension `LET` does that.)

We have to have a restriction of some kind. The SEP-0007 design makes it a compile-time condition (test once per query).

In general, some system have optimizations that on the static analysis of the query. Making the RHS of a LATERAL have runtime variance would interfere.

afs replied:
Sorry, I have to get Jena 5.3.0 out without this.

This issue is important and something that is, I hope, going to get more importance in the SPARQL community.

[WG sparql-query PR 177](https://github.com/w3c/sparql-query/pull/177) at least gets the essential machinery into the resp with the WG contribution licnesing. A small step, but a step.

Aklakan replied:
> In general, some system have optimizations that on the static analysis of the query.
Well, the tradeoff to what extend existing implementations need to drive restrictions of new features in a standard.

For the sake of closing the issue, should I modify/simplify the logic in QueryIterLateral or should I leave the slightly more general approach as it is for now? Essentially, the var scope check should prevent non-empty sets of commonVars.

Aklakan replied:
> Sorry, I have to get Jena 5.3.0 out without this.

No worries, whenever it's ready.

Aklakan commented:
Just to summarize: The main open point for this PR is whether to remove the `Algebra.compatible` check on the `commonVars` under the assumption that the bindings have to be compatible anyway (because of `SyntaxVarScope.checkLATERAL`).

Aklakan commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterLateral.java:
```diff
@@ -291,18 +291,25 @@ public Op transform(OpTable opTable) {
-            binding.vars().forEachRemaining(vars::add);
-            TableN table2 = new TableN(vars);
+
+            TableBuilder tableBuilder = TableFactory.builder();
```
Based on the discussion, I now added a TableBuilder class for building immutable Tables.

Aklakan commented:
I added a TableBuilder class for building immutable tables.
In accordance with the discussion, in `QueryIterLateral` I removed the algebra compatibility check and updated the table creation to use the new TableBuilder. Migrating existing table creation would be a future endeavor.

afs commented:
I'd mostly finished reviewing - does the TableBuilder make a difference?

Let me know when this is ready.

The main issues are to do with the formal description of LATERAL, and not the core in this PR which is faithful to it the formal description. A deeper reworking of execution could so better but it would also change other places.

Functionality before optimization!

Aklakan commented:
> I'd mostly finished reviewing - does the TableBuilder make a difference?

It's not a big difference - It exists now because you suggested that this might be the better approach - and I used it in the fixed code.

Other than that its finished.
